### PR TITLE
Harden access to validation data

### PR DESCRIPTION
**Pull request by westonruter**
## Summary

By default the `amp_validate` capability is mapped to `manage_options`. However, it doesn't have to be. A site could disable access to validation data entirely—even to administrators—using a plugin like:

```php
add_filter(
	'map_meta_cap',
	function ( $caps, $cap ) {
		if ( AMP_Validation_Manager::VALIDATE_CAPABILITY === $cap ) {
			$caps[] = 'do_not_allow';
		}
		return $caps;
	},
	10,
	3
);
```

If I try doing that, some unexpected things happen for new screens in 2.2:

1. The settings screen throws an error. Site Scan should be omitted entirely.
2. Site Scanning is able to proceed in the Onboarding Wizard. Site Scan step should be skipped.
2. Support screen is accessible. It should not be available.

Screen Shot 2021-12-03 at 14 24 05

We need to make sure that if a site is restricting access to any validation data by checking first if the user `AMP_Validation_Manager::has_cap()`. 

## Checklist

- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by delawski** [comment]
@westonruter In ae702a3 you'll find the most important change – a new scanner status is introduced there, `STATUS_SKIPPED`. It is set if there's no `VALIDATE_NONCE` passed from the backend meaning that the current user has no capabilities to do AMP validation.

In e656fbf a minor refactor has been done so that the complexity of the `siteScanReducer` function is within the predefined limit.

Lastly, in e483fbe new E2E tests have been introduced. Thanks to them we're sure that if a user has no validation capability, Site Scan won't be available on the AMP Settings screen, in the Onboarding Wizard and after plugin activation.

I think it's now ready for your review.

---

As a side note, I think that our context provider trees are becoming pretty big and complex, with various dependencies between providers. I'm thinking of reevaluating this approach and maybe looking into a more systematic solution like [React Query](https://react-query.tanstack.com/) that would abstract away data fetching, caching and updating. Since data is stored in a global variable (internally), it can be exposed everywhere in a React app using a set of hooks. This might simplify and decouple the client-side data layer.

File: tests/e2e/config/bootstrap.js [hunk 1]
```diff
@@ -236,9 +236,11 @@ async function createTestData() {
  */
 async function setupThemesAndPlugins() {
 	await installLocalPlugin( 'e2e-tests-demo-plugin' );
+	await installLocalPlugin( 'do-not-allow-amp-validate-capability' );
 
-	// If the demo plugin has been already installed it might be activated, too.
+	// If the plugins have been already installed, they may be activated, too. Try deactivating them, just in case.
 	await deactivatePlugin( 'e2e-tests-demo-plugin' );
+	await deactivatePlugin( 'do-not-allow-amp-validate-capability' );
```
Comment (westonruter) [inline tests/e2e/config/bootstrap.js:243]: Nit: the list of plugins could be in an array, and then they could be looped over.

**Review by westonruter — COMMENTED** [review]
@delawski Thank you for the additional changes! Looks great and now behaves as expected.

**Comment by westonruter** [comment]
Merging without approval since I'm the one who started this PR. @delawski I'm assuming you approve 😄  If you can sanity check my additional changes that would be great.

File: tests/e2e/config/bootstrap.js (hunk 1 above)
Comment (delawski) [reply tests/e2e/config/bootstrap.js:243]: Yeah, I was thinking of the same, though there might be some issues when using `await` inside a loop. Let's take care of that as soon as we need to add another plugin.

**Comment by delawski** [comment]
The changes (and the PR overall) look good to me 👍

**Comment by dhaval-parekh** [comment]
**QA Passed**

After using the plugin with the code given above (in the description).

✅ The "AMP Setting" page is functioning correctly and site scanning is omitted on the page.

✅ Support page, Validated URL and Error Index page is disabled for the user.

image

✅ On the "Onboarding Wizard" step for site scan is removed.

image

**Pull request merged**